modules/core/04-channel/keeper/upgrade.go

```diff
@@ -30,72 +29,40 @@ func (k Keeper) ChanUpgradeInit(ctx sdk.Context, portID string, channelID string
 		return 0, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.OPEN, channel.State)
 	}
 
-	// set the restore channel to the current channel and reassign channel state to INITUPGRADE,
-	// if the channel == proposedUpgradeChannel then fail fast as no upgradable fields have been modified.
-	restoreChannel := channel
-	channel.State = types.INITUPGRADE
-	if reflect.DeepEqual(channel, proposedUpgradeChannel) {
-		return 0, errorsmod.Wrap(types.ErrChannelExists, "existing channel end is identical to proposed upgrade channel end")
-	}
-
-	if !k.connectionKeeper.HasConnection(ctx, proposedUpgradeChannel.ConnectionHops[0]) {
-		return 0, errorsmod.Wrapf(connectiontypes.ErrConnectionNotFound, "failed to retrieve connection: %s", proposedUpgradeChannel.ConnectionHops[0])
-	}
-
-	if proposedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
-		proposedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
-		return 0, errorsmod.Wrap(types.ErrInvalidCounterparty, "counterparty port ID and channel ID cannot be upgraded")
-	}
-
-	if !channel.Ordering.SubsetOf(proposedUpgradeChannel.Ordering) {
-		return 0, errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
-	}
-
-	upgradeSequence = uint64(1)
-	if seq, found := k.GetUpgradeSequence(ctx, portID, channelID); found {
-		upgradeSequence = seq + 1
-	}
-
-	upgradeTimeout := types.UpgradeTimeout{
-		TimeoutHeight:    counterpartyTimeoutHeight,
-		TimeoutTimestamp: counterpartyTimeoutTimestamp,
+	if err := k.validateProposedUpgradeFields(ctx, channel, upgrade.ProposedUpgrade); err != nil {
+		return 0, err
 	}
 
-	k.SetUpgradeRestoreChannel(ctx, portID, channelID, restoreChannel)
-	k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
-	k.SetUpgradeTimeout(ctx, portID, channelID, upgradeTimeout)
+	channel.UpgradeSequence++
+	k.SetChannel(ctx, portID, channelID, channel)
 
-	return upgradeSequence, nil
+	return channel.UpgradeSequence, nil
 }
 
 // WriteUpgradeInitChannel writes a channel which has successfully passed the UpgradeInit handshake step.
 // An event is emitted for the handshake step.
-func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID, proposedVersion string, upgradeSequence uint64, upgradeChannel types.Channel) {
+func (k Keeper) WriteUpgradeInit(ctx sdk.Context, portID, channelID string, upgrade types.Upgrade) {
 	defer telemetry.IncrCounter(1, "ibc", "channel", "upgrade-init")
 
 	channel, found := k.GetChannel(ctx, portID, channelID)
 	if !found {
 		panic(fmt.Sprintf("failed to retrieve channel %s on port %s", channelID, portID))
 	}
 
-	// assign directly the fields that are modifiable.
-	// counterparty fields may not be changed.
 	channel.State = types.INITUPGRADE
-	channel.Version = proposedVersion
-	channel.Ordering = upgradeChannel.Ordering
-	channel.ConnectionHops = upgradeChannel.ConnectionHops
+	k.SetChannel(ctx, portID, channelID, channel)
```

Comment:
I see we are setting the channel in `ChanUpgradeInit` (with a new upgrade sequence) and setting it again here.Is it worthwhile trying to get this down to a single set operation or is it not that significant?